### PR TITLE
Fix count aggregate field in inline fragments

### LINKED ISSUE #2982: Count aggregation breaks when querying fields in inline fragments
**Describe the bug**
When querying the `count` aggregate field on an inline fragment, incorrect cypher syntax is generated and the query fails.

The resulting error is:

```
Received: [[GraphQLError: Variable `count` not defined (line 13, column 117 (offset: 648))
"        WITH this_Posts { __resolveType: "BlogArticle", __id: id(this), CommentsAggregate:  count: this_Posts_CommentsAggregate_var2 } AS this_Posts"
```

**Type definitions**
The following type definitions are used:

```graphql
type User {
    id: ID!
    posts: [Post!]! @relationship(type: "USER_POSTS", direction: OUT)
}

interface Post {
    id: ID!
}

type Comment {
    id: ID!
}

type BlogArticle implements Post {
    id: ID!
    comments: [Comment!]! @relationship(type: "ARTICLE_COMMENTS", direction: OUT)
}
```

**To Reproduce**

1. With the type definitions above, create a `User` and a `BlogArticle` and `MERGE` them
2. Run the following query:

```graphql
query {
    users {
        posts {
            ... on BlogArticle {
                commentsAggregate {
                    count
                }
            }
        }
    }
}
```

3. See the resulting error

**Expected behavior**
The count aggregate field should be returned (with a value of 0 in the case above)

* OS: Linux
* Version: @neo4j/graphql@3.17.0
* Node.js version: 18.14.2

**Additional context**
Please see #2983 for a test case triggering this bug and the fix.

### PR DESCRIPTION
**Pull request by happenslol**
# Description

This PR fixes a syntax error that occurred when querying the count aggregate field inside an inline fragment. The problem here was that the code responsible for sanitizing nested projections in inline fragments would remove all opening and closing brackets in the nested query, which would result in invalid syntax when the nested query contained more nested fields:

`{ .id } -> .id // works`  
`{ fieldAggregate: { count: var }} -> fieldAggregate:  count: var // syntax error`

The offending regex here was `/{|)/gm`. I've looked through all the test cases triggering this code, and it seems there is never more than 1 set of brackets in all other cases, and never any multiline strings. I've adjusted the replace regexes to first remove any newlines, and then clear any opening brackets and closing brackets from the start and end of the string, including surrounding whitespace.

I'm not sure if this is the best way to do it, since we essentially want to remove "1 surrounding pair of brackets" here, but this seemed like the easiest way to achieve this without using any advanced regex features. Please let me know if you have a better idea.

## Complexity

Complexity: Low

# Issue

Closes #2982

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [ ] Documentation has been updated
- [ ] TCK tests have been updated
- [x] Integration tests have been updated
- [ ] Example applications have been updated
- [x] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-stage-2983.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 40 | 50 | 1134 |
|  aggregations.NestedAggregation | 13514 | 13514 | 107 | 138 | 2174 |
|  aggregations.AggregationWithWhere | 10942 | 10942 | 75 | 96 | 2174 |
|  aggregations.AggregationWhereWithinNestedRelationships | 20101954 | 20101954 | 3707 | 3429 | 2008534 |
|  aggregations.AggregationWhereWithinNestedConnections | 20101954 | 20101954 | 3253 | 3283 | 2008534 |
|  aggregations.NestedCountFromMovieToActors | 8694 | 8694 | 80 | 107 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 8900 | 8900 | 64 | 152 | 2174 |
|  aggregations.DeeplyNestedCount | 12062296 | 12062298 | 5681 | 6090 | 2008534 |
|  batch-create.BatchCreate | 3600 | 3600 | 116 | 167 | 600 |
|  connect.createAndConnect | 12419 | 12419 | 231 | 419 | 3003 |
|  connections.Connection | 13042 | 13042 | 131 | 191 | 2174 |
|  connections.NestedConnection | 38237 | 38231 | 135 | 219 | 4516 |
|  connections.ConnectionWithSort | 2277 | 2277 | 85 | 99 | 1040 |
|  connections.ConnectionWithSortAndCypher | 2277 | 2277 | 236 | 393 | 1040 |
|  create.SimpleMutation | 6 | 6 | 38 | 52 | 1 |
|  cypher-directive.TopLevelCypherDirective | 99525 | 99525 | 870 | 1064 | 10168 |
|  cypher-directive.TopLevelCypherDirectiveWithColumnName | 127994 | 127994 | 304 | 330 | 10168 |
|  delete.SimpleDelete | 18361 | 18361 | 281 | 296 | 1040 |
|  delete.NestedDeleteInUpdate | 16779 | 16779 | 307 | 309 | 2040 |
|  query.SimpleQuery | 14120 | 14120 | 207 | 137 | 2174 |
|  query.QueryWhere | 8675 | 8675 | 46 | 52 | 2166 |
|  query.SimpleQueryWithNestedWhere | 8846 | 8846 | 77 | 82 | 2166 |
|  query.Nested | 10084290 | 10084290 | 11897 | 12350 | 2008534 |
|  query.NestedWithFilter | 37172 | 37172 | 169 | 173 | 4511 |
|  query.OrFilterOnRelationships | 37426 | 35944 | 220 | 240 | 2097 |
|  query.OrFilterOnRelationshipsAndNested | 31477 | 30014 | 253 | 325 | 2097 |
|  query.QueryWithNestedIn | 12719 | 12599 | 96 | 73 | 1341 |
|  query.NestedConnectionWhere | 8794 | 8794 | 78 | 83 | 2174 |
|  query.DeeplyNestedConnectionWhere | 9004 | 9004 | 110 | 133 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 6251 | 6251 | 270 | 207 | 1134 |
|  query.NestedWithRelationshipSingleFilters | 3021189 | 3021189 | 810 | 855 | 1003942 |
|  query.Fulltext | 96 | 96 | 45 | 39 | 16 |
|  query.FulltextWithNestedQuery | 571 | 571 | 65 | 58 | 84 |
|  sorting.SortMultipleTypes | 2493 | 2493 | 94 | 88 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 1439 | 1439 | 239 | 316 | 1040 |
|  sorting.SortOnNestedFields | 13042 | 13042 | 92 | 79 | 2174 |
|  sorting.SortDeeplyNestedFields | 38965 | 38965 | 149 | 167 | 4516 |
|  sorting.SortWithTopLevelCypher | 2119 | 2119 | 143 | 223 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 85 | 104 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 84 | 125 | 35 |
|  unions.NestedUnion | 398553 | 398553 | 397 | 425 | 33033 |
|  unions.NestedUnionWithMissingFields | 372527 | 372527 | 411 | 373 | 33033 |
|  update.NestedUpdate | 2119 | 2119 | 52 | 64 | 1040 |

</details>
        
Old Schema Generation: 40.133s
Schema Generation: 40.729s

File: packages/graphql/src/translate/create-projection-and-params.ts [hunk 1]
```diff
@@ -185,7 +185,14 @@ export default function createProjectionAndParams({
...
 
                     const nestedProjection = new Cypher.RawCypher((env) => {
-                        const nestedProj = recurse.projection.getCypher(env).replace(/{|}/gm, "").trim();
+                        // The nested projection will be surrounded by brackets, so we want to remove
+                        // any linebreaks, and then the first opening and the last closing bracket of the line,
+                        // as well as any surrounding whitespace.
+                        const nestedProj = recurse.projection
+                            .getCypher(env)
+                            .replaceAll("\n", " ")
+                            .replace(/^\s*{\s*|\s*}\s*$/g, "");
```
Comment (angrykoala) [inline packages/graphql/src/translate/create-projection-and-params.ts:194]: `\s` already takes white space into account, so the first step of replacing `\n` into `" "` should not be needed (unless there is another reason for this addition that I have not though of :thinking:?

Additionally, I would consider either changing this to `replaceAll` (as it forces the regex to be global and better denotes what we try to do). Whatever you feel is nicer

 ```suggestion
                             .replaceAll(/^\s*{\s*|\s*}\s*$/g, "");
```

Another suggestion on this regex, could be adding parenthesis to denote the separate alternatives:

```regex
(^\s*{\s*)|(\s*}\s*$)
```

This is mainly for readability, and making this regex less prone to errors if modified in the future, but I'm happy with your current regex as well if you feel this is not a good idea

File: packages/graphql/tests/integration/issues/2982.int.test.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+import { graphql } from "graphql";
+import type { Driver, Session } from "neo4j-driver";
+import { generate } from "randomstring";
+import { UniqueType } from "../../utils/graphql-types";
+import { Neo4jGraphQL } from "../../../src/classes";
+import Neo4j from "../neo4j";
+import { cleanNodes } from "../../utils/clean-nodes";
+
+describe("https://github.com/neo4j/graphql/issues/2982", () => {
```
Comment (angrykoala) [inline packages/graphql/tests/integration/issues/2982.int.test.ts:28]: :clap: Nice test

**Review by angrykoala — CHANGES_REQUESTED** [review]

**Comment by angrykoala** [comment]
Thanks for your contribution, it looks good, just a proposal on the generated regex and it should be good to go

File: packages/graphql/src/translate/create-projection-and-params.ts (hunk 1 above)
Comment (angrykoala) [reply packages/graphql/src/translate/create-projection-and-params.ts]: Hi, if I'm not mistaken, that would only happen if the multiline option (`/m`) is set in the regex, which is not the case so it should be safe to let the regex take care of that

Comment (happenslol) [reply packages/graphql/src/translate/create-projection-and-params.ts]: Just checked and you're completely right, my bad. Removing that line and pushing again.

Comment (angrykoala) [reply packages/graphql/src/translate/create-projection-and-params.ts]: no worries, thanks :smile:

**Comment by happenslol** [comment]
Seems like one of the actions is failing now, I'm not quite sure why though. Anything else you need from me to get this going?

**Comment by angrykoala** [comment]
Hi @happenslol 
The failing build seems to be unrelated to your PR so it should be good. Everything looks good in your code.

I think you haven't enabled "Allow changes by maintainers" in this PR. Please enable it so I can update the changelog before merging. That should be the only thing left :smile:

**Comment by happenslol** [comment]
@angrykoala Since the fork was made on my company's organization, I don't have the option to allow edits sadly :-/ If you tell me what to add, I can add the changeset myself though.

**Comment by angrykoala** [comment]
Sure, just run yarn changeset, select the graphql package and a minor version and write a message regarding the fix

**Comment by happenslol** [comment]
All done!

**Comment by angrykoala** [comment]
Thanks @happenslol !

**Pull request merged**